DR_0008_role-based-permissions.md

```diff
@@ -0,0 +1,187 @@
+
+```python
+project_user = Group(name=’ProjectUser’, permissions=[p1,...,pn])
+```
+
+##### Roles
+
+We define an abstract base model Role with the following attributes:
+
+```python
+class Role (model.Model):
+	class Meta:
+	    abstract = True
+		user = ForeignKey(‘accounts.User’)
+
+	@cached_property
+	def permissions(self):
+		return [perm.codename
+				for perm in self.group.permissions.all()]
+```
+
+We then modify `OrganizationRole` and `ProjectRole` to extend `Role`:
+
+```python
+class OrganizationRole(Role):
+	organization = ForeignKey(Organization)
+	group = models.ForeignKey(
+        'auth.Group', related_name='organization_roles')
+
+class ProjectRole(Role):
+	project = ForeignKey(Project)
+	role = CharField(choices=ROLE_CHOICES, default=’PU’)
+	group = models.ForeignKey(
+        'auth.Group', related_name='project_roles')
+```
+
+
+When a Role is created it is associated with its corresponding Group, eg:
+
+```python
+pu = Group.objects.get(name='ProjectUser')
+role = ProjectRole(user=user, group=pu,
+				   project=project, role=’PU’)
+```
+When a permissions check need to be made to see if a user can perform a particular action the action will be allowed if the permission is in `role.permissions`.
+
+For the original discussions around this approach see https://docs.google.com/document/d/1CfuaHQ8Sg-EeNp-JeGB_YcFpei7G3aBkpNy4yYB2YHk/edit#
+
+
+## Database changes
+
+The proposed solution will introduce a number of database changes. Both `OrganizationRole` and `ProjectRole` will need to be updated with a `Group` attribute which will hold the appropriate role group and the associated set of `Permissions`. In addition, the `admin` field on the `OrganizationRole` model will be removed from the database and replaced with a boolean property. The property will be `True` if the `OrganizationRole` belongs to the `OrgAdmin` group. Migrations will be required to add the `group` field to the appropriate models, to set this field to `not-null`, to update existing `OrganizationRole` and `ProjectRole` instances with the corresponding groups.  Ideally, test cases will be provided to verify the correctness of these proposed changes.
+
+## Security Considerations
+
+Since this proposed solution will determine access to platform data in future, it will need to be thoroughly tested.
+
+## Scalablity
+
+
+The current permissions system implementation based on `django-tutelary` is not scalable, see analysis in the backlog issue [here](https://github.com/Cadasta/cadasta-platform/issues/1467). However, the current proposed solution still requires iterating over organizations and projects for per-object permissions checking - particularly in list views. The lack of bulk test data makes it difficult to test scalability limits. The implementation should be tested against a mirror of the production database to judge whether there is an improvement in performance v's tutelary.
```

Comment:
I recommend testing against a copy of the demo database as well as against the production copy. There are a lot more records in demo, including user accounts and organizations/projects that have been problematic under tutelary (whereas production records haven't yet triggered these performance issues).